src/amo/reducers/home.js

```diff
@@ -281,6 +286,15 @@ const reducer = (
       };
     }
 
+    case LOCATION_CHANGE: {
+      const newState = state.clientSideLocationChanges ? initialState : state;
```

Comment:
I find this change complicated to read and given that it is tied to the routing layer, I think it should be a bit more readable. Here is my suggestion:

```diff
diff --git a/src/amo/reducers/home.js b/src/amo/reducers/home.js
index 97e9102c5..a7ebbe985 100644
--- a/src/amo/reducers/home.js
+++ b/src/amo/reducers/home.js
@@ -108,7 +108,7 @@ export type HeroShelvesType = {|
 export type HomeState = {
   collections: Array<Object | null>,
   heroShelves: HeroShelvesType | null,
-  clientSideLocationChanges: boolean,
+  resetStateOnNextChange: boolean,
   resultsLoaded: boolean,
   shelves: { [shelfName: string]: Array<AddonType> | null },
 };
@@ -116,7 +116,7 @@ export type HomeState = {
 export const initialState: HomeState = {
   collections: [],
   heroShelves: null,
-  clientSideLocationChanges: false,
+  resetStateOnNextChange: false,
   resultsLoaded: false,
   shelves: {},
 };
@@ -286,12 +286,23 @@ const reducer = (
       };
     }
 
+    // See: https://github.com/mozilla/addons-frontend/issues/8601
     case LOCATION_CHANGE: {
-      const newState = state.clientSideLocationChanges ? initialState : state;
+      if (_config.get('server')) {
+        // We only focus on client side navigation.
+        return state;
+      }
+
+      // When the client initializes, it updates its location. On next location
+      // change, we want to reset this state to fetch fresh data once user goes
+      // back to the homepage.
+      if (state.resetStateOnNextChange) {
+        return initialState;
+      }
 
       return {
-        ...newState,
-        clientSideLocationChanges: !_config.get('server'),
+        ...state,
+        resetStateOnNextChange: true,
       };
     } 
```

WDYT?

Reply:
This would help me comprehend the code better as well.

Reply:
Yes, I like this suggestion. It does make the code a bit easier to reason about and the name is better. Thanks @willdurand.